File: Algorithms/Python3.x/1195-Fizz_Buzz_Multithreaded.py
# Runtime: 48 ms, faster than 89.09% of Python3 online submissions for Fizz Buzz Multithreaded.
# Memory Usage: 16.2 MB, less than 100.00% of Python3 online submissions for Fizz Buzz Multithreaded.
import threading


class FizzBuzz:
    def __init__(self, n: int):
        self.n = n + 1
        self.f = threading.Semaphore(0)
        self.b = threading.Semaphore(0)
        self.fb = threading.Semaphore(0)
        self.nn = threading.Semaphore(1)
        self.cur = 1

    # printFizz() outputs "fizz"
    def fizz(self, printFizz: 'Callable[[], None]') -> None:
        for i in range(1, self.n):
            if i % 3 == 0 and i % 5:
                self.f.acquire()
                printFizz()
                self.cur += 1
                if self.cur % 5 == 0:
                    self.b.release()
                else:
                    self.nn.release()

    # printBuzz() outputs "buzz"
    def buzz(self, printBuzz: 'Callable[[], None]') -> None:
        for i in range(1, self.n):
            if i % 5 == 0 and i % 3:
                self.b.acquire()
                printBuzz()
                self.cur += 1
                if self.cur % 3 == 0:
                    self.f.release()
                else:
                    self.nn.release()

    # printFizzBuzz() outputs "fizzbuzz"
    def fizzbuzz(self, printFizzBuzz: 'Callable[[], None]') -> None:
        for i in range(1, self.n):
            if i % 5 == 0 and i % 3 == 0:
                self.fb.acquire()
                printFizzBuzz()
                self.cur += 1
                self.nn.release()

    # printNumber(x) outputs "x", where x is an integer.
    def number(self, printNumber: 'Callable[[int], None]') -> None:
        for i in range(1, self.n):
            if i % 5 and i % 3:
                self.nn.acquire()
                printNumber(i)
                self.cur += 1
                if self.cur % 3 == 0 and self.cur % 5 == 0:
                    self.fb.release()  # fb add 1
                elif self.cur % 3 == 0:
                    self.f.release()  # f add 1
                elif self.cur % 5 == 0:
                    self.b.release()  # b add 1
                else:
                    self.nn.release()

File: Algorithms/C++/1195-Fizz_Buzz_Multithreaded.cpp
// Runtime: 28 ms, faster than 94.35% of C++ online submissions for Fizz Buzz Multithreaded.
// Memory Usage: 9.2 MB, less than 100.00% of C++ online submissions for Fizz Buzz Multithreaded.
// condition_variable + mutex
class FizzBuzz {
private:
    int n;
    int cnt;
    mutex m;
    condition_variable cv;

public:
    FizzBuzz(int n) {
        this->n = n;
        this->cnt = 1;
    }

    // printFizz() outputs "fizz".
    void fizz(function<void()> printFizz) {
        while (true) {
            unique_lock<mutex> lock(m);
            while (cnt <= n && (cnt % 3 != 0 || cnt % 5 == 0))
                cv.wait(lock);
            if (cnt > n)
                return ;
            printFizz();
            ++cnt;
            cv.notify_all();
        }
    }

    // printBuzz() outputs "buzz".
    void buzz(function<void()> printBuzz) {
        while (true) {
            unique_lock<mutex> lock(m);
            while (cnt <= n && (cnt % 5 != 0 || cnt % 3 == 0))
                cv.wait(lock);
            if (cnt > n)
                return ;
            printBuzz();
            ++cnt;
            cv.notify_all();
        }
    }

    // printFizzBuzz() outputs "fizzbuzz".
	void fizzbuzz(function<void()> printFizzBuzz) {
        while (true) {
            unique_lock<mutex> lock(m);
            while (cnt <= n && (cnt % 5 != 0 || cnt % 3 != 0))
                cv.wait(lock);
            if (cnt > n)
                return ;
            printFizzBuzz();
            ++cnt;
            cv.notify_all();
        }
    }

    // printNumber(x) outputs "x", where x is an integer.
    void number(function<void(int)> printNumber) {
        while (true) {
            unique_lock<mutex> lock(m);
            while (cnt <= n && (cnt % 5 == 0 || cnt % 3 == 0))
                cv.wait(lock);
            if (cnt > n)
                return ;
            printNumber(cnt);
            ++cnt;
            cv.notify_all();
        }
    }
};
